sdk/go/common/util/env/env.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A small library for creating consistent and documented environmental variable accesses.
+//
+// Public environmental variables should be declared as a module level variable.
+package env
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+var Experimental = Bool("EXPERIMENTAL", "Enable experimental options and commands")
+
+type Var struct {
+	name        string
+	Value       Value
+	Description string
+	options     options
+}
+
+const Prefix = "PULUMI"
+
+func (v Var) Name() string {
+	return v.options.name(v.name)
+}
+
+func (v Var) Requires() []BoolValue { return v.options.prerequs }
+
+var envVars []Var
+
+// Variables is a list of variables declared.
+func Variables() []Var {
+	return envVars
+}
+
+type Option func(*options)
+
+type options struct {
+	prerequs []BoolValue
+	noPrefix bool
+}
+
+func (o options) name(underlying string) string {
+	if o.noPrefix {
+		return underlying
+	}
+	return Prefix + "_" + underlying
+}
+
+func Needs(val BoolValue) Option {
+	return func(o *options) {
+		o.prerequs = append(o.prerequs, val)
+	}
+}
+
+func NoPrefix(opts *options) {
+	opts.noPrefix = true
+}
+
+type Value interface {
+	fmt.Stringer
+	Underlying() (string, bool)
+	Var() Var
+
+	setVar(Var)
+	formattedValue() string
+}
+
+type value struct {
+	underlying string
+	present    bool
+	_var       Var
```

Comment:
You can use an anon field here if you want. Having the underscore seems to violate Go style conversions, especially since `_var` is used.

```suggestion
	Var
```